File: daemon/internal/vault/vault.go
// Package vault — Cryptomator-compatible vault (format 8, SIV_GCM).
// Reads real Cryptomator vaults created by the official app.
package vault

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/google/uuid"
	"golang.org/x/crypto/scrypt"
)

// ErrWrongPassword is returned when the password is incorrect (AES-KW integrity check failed).
var ErrWrongPassword = errors.New("wrong vault password")

// Vault is an unlocked Cryptomator vault (format 8, SIV_GCM).
type Vault struct {
	encKey []byte // 32B — AES-GCM content encryption + header key
	macKey []byte // 32B — SIV MAC half + JWT signature
}

// masterkeyFile is the JSON structure of masterkey.cryptomator.
type masterkeyFile struct {
	Version          int    `json:"version"`
	ScryptSalt       string `json:"scryptSalt"`
	ScryptCostParam  int    `json:"scryptCostParam"`
	ScryptBlockSize  int    `json:"scryptBlockSize"`
	PrimaryMasterKey string `json:"primaryMasterKey"`
	HmacMasterKey    string `json:"hmacMasterKey"`
	VersionMac       string `json:"versionMac"`
}

// Open opens a Cryptomator vault at vaultDir using the given password.
// Reads masterkey.cryptomator and vault.cryptomator from the local filesystem.
// Wrong password → ErrWrongPassword.
func Open(vaultDir, password string) (*Vault, error) {
	mkData, err := os.ReadFile(filepath.Join(vaultDir, "masterkey.cryptomator"))
	if err != nil {
		return nil, fmt.Errorf("vault: reading masterkey.cryptomator: %w", err)
	}
	jwtData, err := os.ReadFile(filepath.Join(vaultDir, "vault.cryptomator"))
	if err != nil {
		return nil, fmt.Errorf("vault: reading vault.cryptomator: %w", err)
	}
	return deriveVault(mkData, jwtData, password)
}

// OpenWithKeys opens a vault directly from master keys recovered from a recovery phrase
// (see RecoveryToKeys), bypassing the password. It verifies the keys against
// vault.cryptomator, so a phrase from a different vault is rejected.
func OpenWithKeys(vaultDir string, encKey, macKey []byte) (*Vault, error) {
	jwtData, err := os.ReadFile(filepath.Join(vaultDir, "vault.cryptomator"))
	if err != nil {
		return nil, fmt.Errorf("vault: reading vault.cryptomator: %w", err)
	}
	if err := verifyVaultJWT(strings.TrimSpace(string(jwtData)), encKey, macKey); err != nil {
		return nil, fmt.Errorf("vault: recovery phrase does not match this vault: %w", err)
	}
	return &Vault{encKey: encKey, macKey: macKey}, nil
}

// deriveVault unwraps the master keys from masterkey.cryptomator bytes and verifies
// vault.cryptomator bytes. Wrong password → ErrWrongPassword.
func deriveVault(mkData, jwtData []byte, password string) (*Vault, error) {
	// 1. Parse masterkey.cryptomator
	var mk masterkeyFile
	if err := json.Unmarshal(mkData, &mk); err != nil {
		return nil, fmt.Errorf("vault: parsing masterkey.cryptomator: %w", err)
	}

	// 2. Decode salt and wrapped keys
	salt, err := base64.StdEncoding.DecodeString(mk.ScryptSalt)
	if err != nil {
		return nil, fmt.Errorf("vault: decoding scryptSalt: %w", err)
	}
	wrappedEnc, err := base64.StdEncoding.DecodeString(mk.PrimaryMasterKey)
	if err != nil {
		return nil, fmt.Errorf("vault: decoding primaryMasterKey: %w", err)
	}
	wrappedMac, err := base64.StdEncoding.DecodeString(mk.HmacMasterKey)
	if err != nil {
		return nil, fmt.Errorf("vault: decoding hmacMasterKey: %w", err)
	}

	// 3. Derive KEK via scrypt
	kek, err := scrypt.Key([]byte(password), salt, mk.ScryptCostParam, mk.ScryptBlockSize, 1, 32)
	if err != nil {
		return nil, fmt.Errorf("vault: scrypt: %w", err)
	}

	// 4. Unwrap keys (AES-KW). Wrong password → IV mismatch
	encKey, err := aesKWUnwrap(kek, wrappedEnc)
	if err != nil {
		return nil, ErrWrongPassword
	}
	macKey, err := aesKWUnwrap(kek, wrappedMac)
	if err != nil {
		return nil, ErrWrongPassword
	}

	// 5. Verify vault.cryptomator (compact JWT, alg=HS256)
	if err := verifyVaultJWT(strings.TrimSpace(string(jwtData)), encKey, macKey); err != nil {
		return nil, fmt.Errorf("vault: verifying vault.cryptomator: %w", err)
	}

	return &Vault{encKey: encKey, macKey: macKey}, nil
}

// Create generates a new Cryptomator vault at vaultDir: encKey/macKey (CSPRNG), writes
// masterkey.cryptomator (scrypt KEK + AES-KW wrap) and vault.cryptomator (JWT HS256).
// Returns the opened Vault.
func Create(vaultDir, password string) (*Vault, error) {
	// 1. Generate keys
	encKey := make([]byte, 32)
	macKey := make([]byte, 32)
	scryptSalt := make([]byte, 8)
	if _, err := rand.Read(encKey); err != nil {
		return nil, fmt.Errorf("vault: generating encKey: %w", err)
	}
	if _, err := rand.Read(macKey); err != nil {
		return nil, fmt.Errorf("vault: generating macKey: %w", err)
	}
	if _, err := rand.Read(scryptSalt); err != nil {
		return nil, fmt.Errorf("vault: generating scryptSalt: %w", err)
	}

	// 2. KEK = scrypt(password, salt, N=32768, r=8, p=1, dkLen=32)
	kek, err := scrypt.Key([]byte(password), scryptSalt, 32768, 8, 1, 32)
	if err != nil {
		return nil, fmt.Errorf("vault: scrypt: %w", err)
	}

	// 3. Wrap keys (AES-KW)
	wrappedEnc, err := aesKWWrap(kek, encKey)
	if err != nil {
		return nil, fmt.Errorf("vault: wrap encKey: %w", err)
	}
	wrappedMac, err := aesKWWrap(kek, macKey)
	if err != nil {
		return nil, fmt.Errorf("vault: wrap macKey: %w", err)
	}

	// 4. versionMac = HMAC-SHA256(macKey, BE32(999))
	vm := vaultVersionMac(macKey, 999)

	// 5. Write masterkey.cryptomator (field order matches the reference)
	mk := masterkeyFile{
		Version:          999,
		ScryptSalt:       base64.StdEncoding.EncodeToString(scryptSalt),
		ScryptCostParam:  32768,
		ScryptBlockSize:  8,
		PrimaryMasterKey: base64.StdEncoding.EncodeToString(wrappedEnc),
		HmacMasterKey:    base64.StdEncoding.EncodeToString(wrappedMac),
		VersionMac:       base64.StdEncoding.EncodeToString(vm),
	}
	mkBytes, err := json.MarshalIndent(mk, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("vault: serializing masterkey: %w", err)
	}
	if err := os.MkdirAll(vaultDir, 0o755); err != nil {
		return nil, fmt.Errorf("vault: creating directory: %w", err)
	}
	if err := os.WriteFile(filepath.Join(vaultDir, "masterkey.cryptomator"), mkBytes, 0o600); err != nil {
		return nil, fmt.Errorf("vault: writing masterkey.cryptomator: %w", err)
	}

	// 6. Write vault.cryptomator (compact JWT, HS256, key = encKey||macKey)
	jwtToken, err := buildVaultJWT(encKey, macKey)
	if err != nil {
		return nil, fmt.Errorf("vault: building vault.cryptomator JWT: %w", err)
	}
	if err := os.WriteFile(filepath.Join(vaultDir, "vault.cryptomator"), []byte(jwtToken), 0o644); err != nil {
		return nil, fmt.Errorf("vault: writing vault.cryptomator: %w", err)
	}

	return &Vault{encKey: encKey, macKey: macKey}, nil
}

// vaultVersionMac computes HMAC-SHA256(macKey, BE32(version)).
func vaultVersionMac(macKey []byte, version uint32) []byte {
	h := hmac.New(sha256.New, macKey)
	var vb [4]byte
	binary.BigEndian.PutUint32(vb[:], version)
	h.Write(vb[:])
	return h.Sum(nil)
}

// buildVaultJWT builds a compact JWT for vault.cryptomator.
// Header: {"kid":"masterkeyfile:masterkey.cryptomator","alg":"HS256","typ":"JWT"}
// Payload: {"jti":"<uuid>","format":8,"cipherCombo":"SIV_GCM","shorteningThreshold":220}
// Signature: HS256, key = encKey||macKey (64B).
func buildVaultJWT(encKey, macKey []byte) (string, error) {
	headerJSON := `{"kid":"masterkeyfile:masterkey.cryptomator","alg":"HS256","typ":"JWT"}`
	payloadJSON := fmt.Sprintf(`{"jti":%q,"format":8,"cipherCombo":"SIV_GCM","shorteningThreshold":220}`,
		uuid.NewString())

	headerPart := base64.RawURLEncoding.EncodeToString([]byte(headerJSON))
	payloadPart := base64.RawURLEncoding.EncodeToString([]byte(payloadJSON))

	sigInput := headerPart + "." + payloadPart
	sigKey := append(encKey, macKey...)
	h := hmac.New(sha256.New, sigKey)
	h.Write([]byte(sigInput))
	sig := h.Sum(nil)

	return sigInput + "." + base64.RawURLEncoding.EncodeToString(sig), nil
}

// verifyVaultJWT verifies the JWT signature and checks that format=8 and cipherCombo=SIV_GCM.
func verifyVaultJWT(token string, encKey, macKey []byte) error {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return errors.New("malformed JWT: expected 3 parts")
	}
	// Signature key = encKey ‖ macKey (64B)
	sigKey := append(encKey, macKey...)

	// Verify HS256 signature: HMAC-SHA256(header.payload)
	h := hmac.New(sha256.New, sigKey)
	h.Write([]byte(parts[0] + "." + parts[1]))
	expectedSig := h.Sum(nil)

	gotSig, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return fmt.Errorf("decoding JWT signature: %w", err)
	}
	if !hmac.Equal(gotSig, expectedSig) {
		return errors.New("JWT signature mismatch — masterkey does not match vault")
	}

	// Parse payload
	payloadBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return fmt.Errorf("decoding JWT payload: %w", err)
	}
	var payload struct {
		Format      int    `json:"format"`
		CipherCombo string `json:"cipherCombo"`
	}
	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
		return fmt.Errorf("parsing JWT payload: %w", err)
	}
	if payload.Format != 8 {
		return fmt.Errorf("unsupported vault format: %d (expected 8)", payload.Format)
	}
	if payload.CipherCombo != "SIV_GCM" {
		return fmt.Errorf("unsupported cipherCombo: %q (expected SIV_GCM)", payload.CipherCombo)
	}
	return nil
}

File: clients/DiscoKit/Sources/DiscoKit/Vault/Vault.swift
import Foundation
import CryptoKit

// Cryptomator-compatible vault (format 8, SIV_GCM). Port of daemon/internal/vault.
// Immutable (keys only) → Sendable.
public final class Vault: Sendable {
    let encKey: [UInt8]   // 32B — AES-GCM content key + header key
    let macKey: [UInt8]   // 32B — SIV MAC half + JWT signature key

    init(encKey: [UInt8], macKey: [UInt8]) {
        self.encKey = encKey
        self.macKey = macKey
    }

    public enum VaultError: Error, Equatable {
        case wrongPassword, badMasterkey, badVaultFile, unsupported(String)
    }

    private struct MasterkeyFile: Decodable {
        let scryptSalt: String
        let scryptCostParam: Int
        let scryptBlockSize: Int
        let primaryMasterKey: String
        let hmacMasterKey: String
    }

    // Open a vault from a local directory using a password.
    public static func open(directory: URL, password: String) throws -> Vault {
        let mkData = try Data(contentsOf: directory.appendingPathComponent("masterkey.cryptomator"))
        let jwt = try String(contentsOf: directory.appendingPathComponent("vault.cryptomator"), encoding: .utf8)
        return try open(masterkeyData: mkData, vaultJWT: jwt, password: password)
    }

    // Open a vault via a source (server): downloads masterkey.cryptomator + vault.cryptomator.
    public static func open(source: VaultFileSource, password: String) async throws -> Vault {
        let mkData = try await source.read("masterkey.cryptomator")
        let jwt = String(decoding: try await source.read("vault.cryptomator"), as: UTF8.self)
        return try open(masterkeyData: mkData, vaultJWT: jwt, password: password)
    }

    static func open(masterkeyData mkData: Data, vaultJWT: String, password: String) throws -> Vault {
        let mk = try JSONDecoder().decode(MasterkeyFile.self, from: mkData)
        guard let salt = Data(base64Encoded: mk.scryptSalt),
              let wrappedEnc = Data(base64Encoded: mk.primaryMasterKey),
              let wrappedMac = Data(base64Encoded: mk.hmacMasterKey) else { throw VaultError.badMasterkey }
        let kek = Scrypt.derive(password: Array(password.utf8), salt: Array(salt),
                                n: mk.scryptCostParam, r: mk.scryptBlockSize, p: 1, dkLen: 32)
        let encKey: [UInt8], macKey: [UInt8]
        do {
            encKey = try AESKeyWrap.unwrap(kek: kek, wrapped: Array(wrappedEnc))
            macKey = try AESKeyWrap.unwrap(kek: kek, wrapped: Array(wrappedMac))
        } catch {
            throw VaultError.wrongPassword
        }
        try verifyVaultJWT(vaultJWT.trimmingCharacters(in: .whitespacesAndNewlines), encKey: encKey, macKey: macKey)
        return Vault(encKey: encKey, macKey: macKey)
    }

    private struct JWTPayload: Decodable { let format: Int; let cipherCombo: String }

    static func verifyVaultJWT(_ token: String, encKey: [UInt8], macKey: [UInt8]) throws {
        let parts = token.split(separator: ".", omittingEmptySubsequences: false).map(String.init)
        guard parts.count == 3 else { throw VaultError.badVaultFile }
        let sigKey = SymmetricKey(data: Data(encKey + macKey))
        let expected = HMAC<SHA256>.authenticationCode(for: Data((parts[0] + "." + parts[1]).utf8), using: sigKey)
        guard let gotSig = base64URLDecode(parts[2]) else { throw VaultError.badVaultFile }
        guard Array(expected) == gotSig else { throw VaultError.wrongPassword }
        guard let payloadData = base64URLDecode(parts[1]),
              let payload = try? JSONDecoder().decode(JWTPayload.self, from: Data(payloadData)) else {
            throw VaultError.badVaultFile
        }
        guard payload.format == 8 else { throw VaultError.unsupported("format \(payload.format)") }
        guard payload.cipherCombo == "SIV_GCM" else { throw VaultError.unsupported(payload.cipherCombo) }
    }
}

// base64url → bytes (with or without padding).
func base64URLDecode(_ s: String) -> [UInt8]? {
    var str = s.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
    while str.count % 4 != 0 { str.append("=") }
    return Data(base64Encoded: str).map(Array.init)
}
